### PR TITLE
Allow downgrading of JSR292 methods

### PR DESCRIPTION
**Pull request by mpirvu**
To improve start-up/ramp-up of applications, OpenJ9 allows some compilations
to be downgraded from "warm" (the default optimization level) to "cold"
optimization level. However, current logic prevents downgrading of methods
that are flagged as MethodHandleInvokes because inlining is crucial for
the performance of such methods.
In order to improve start-up time of applications, this commit allows the
downgrading of JSR292 methods under certain conditions:
1. The JVM is in the start-up phase
2. Shared classes are enabled explicitly on the command line
3. AOT is enabled

The reason is that, if the shared classes are explicitly enabled on the
command line and AOT is not disabled, there is a good likelihood that the
user cares about start-up time.
Performance evaluation has shown that start-up time of Liberty+DT7 can be
improved by 5-6% with Java17 and by ~1% with Java11. There was no throughput
impact from these changes.

Signed-off-by: Marius Pirvu <mpirvu@ca.ibm.com>

### CODE REVIEW

**Comment by mpirvu** [comment]
FYI @vijaysun-omr

**Comment by mpirvu** [comment]
jenkins test sanity all jdk17

File: runtime/compiler/control/CompilationThread.cpp [hunk 1]
```diff
@@ -623,13 +623,16 @@ bool TR::CompilationInfo::shouldDowngradeCompReq(TR_MethodToBeCompiled *entry)
...
-      else if (fe->isThunkArchetype(method))
+      // Don't downgrade if method is JSR292 because inlining in those is vital. See CMVC 200145
+      // However, during start-up phase, allow such downgrades if SCC has been specified on the command line
+      // (the check for J9SHR_RUNTIMEFLAG_ENABLE_CACHE_NON_BOOT_CLASSES prevents the downgrading when
+      // the default SCC is used)
+      if (((_J9ROMMETHOD_J9MODIFIER_IS_SET(romMethod, J9AccMethodHasMethodHandleInvokes)) || fe->isThunkArchetype(method)) &&
+           (_jitConfig->javaVM->phase == J9VM_PHASE_NOT_STARTUP ||
+            !TR::Options::sharedClassCache() ||
+            !J9_ARE_ALL_BITS_SET(jitConfig->javaVM->sharedClassConfig->runtimeFlags, J9SHR_RUNTIMEFLAG_ENABLE_CACHE_NON_BOOT_CLASSES)
+           )
```
Comment (vijaysun-omr) [inline runtime/compiler/control/CompilationThread.cpp:634]: The comment talks about `3. AOT` is enabled but this condition is not testing for that ?

Comment (mpirvu) [reply runtime/compiler/control/CompilationThread.cpp:634]: `TR::Options::sharedClassCache()` only returns true it AOT is enabled

**Review by vijaysun-omr — APPROVED** [review]

**Comment by vijaysun-omr** [comment]
Simple heuristic change. Checks have all passed. Merging.

**Pull request merged**